File: server/charge_test.go
package main

import "testing"

func TestElectricChargeLifecycle(t *testing.T) {
	s := initial()
	s.Controlled[1] = 16
	s.Players[16].X = 0
	s.Players[16].Z = 0
	s.Ball = Ball{H: .75, VX: 1, Charged: true, Electric: 1, ElectricBudget: 1, LastTouch: 7, Owner: -1}
	s.catchBall()
	if s.Ball.Owner != -1 || s.Ball.Electric != 0 || !s.Ball.Charged {
		t.Fatal("last hit removed charged state")
	}
	s.Players[16].Stun = 10
	s.Controlled[0] = 7
	s.Players[7].X = 0
	s.Players[7].Z = 0
	s.catchBall()
	if s.Ball.Owner != 7 || !s.Ball.Charged {
		t.Fatal("friendly catch cleared charge")
	}
	s.throw(7, false)
	if s.Ball.Charged || s.Ball.ElectricBudget != 1 {
		t.Fatal("throw did not reset electric state")
	}
	for _, vx := range []float64{0, 1} {
		s = initial()
		s.Controlled[1] = 16
		s.Players[16].X = 0
		s.Players[16].Z = 0
		s.Ball = Ball{H: .75, VX: vx, Charged: true, LastTouch: 7, Owner: -1}
		s.catchBall()
		if s.Ball.Owner != 16 || s.Ball.Charged != (vx == 0) {
			t.Fatal("opposing catch charge state", vx)
		}
	}
}

func TestElectroballFallDirection(t *testing.T) {
	for _, dir := range [][2]float64{{1, 0}, {1, 1}, {0, 1}, {-1, 1}, {-1, 0}, {-1, -1}, {0, -1}, {1, -1}} {
		s := initial()
		s.Controlled[1] = 16
		p := &s.Players[16]
		p.X, p.Z = 0, 0
		s.Players[7].FX, s.Players[7].FZ = -dir[0], -dir[1]
		vx, vz := dir[0]*8, dir[1]*8
		if vx == 0 {
			vx = 4
		}
		if vz == 0 {
			vz = 4
		}
		s.Ball = Ball{Owner: -1, H: .75, DirX: dir[0], DirZ: dir[1], VX: vx, VZ: vz, Charged: true, Electric: 1, ElectricBudget: 1, LastTouch: 7}
		s.catchBall()
		if p.Action != 4 || p.Stun != 26./25 || p.fallX != dir[0]*3*velocityUnit || p.fallZ != dir[1]*3*velocityUnit {
			t.Fatal("electroball fall", dir, p)
		}
		if s.Ball.Electric != 0 || s.Ball.Owner != -1 {
			t.Fatal("charge consumption")
		}
	}
}

File: src/charge.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { initial, catchBall, throwBall } from "./game.js";
test("depleted electroball remains charged, friendly catch retains it, throw clears it", () => {
  const s = initial();
  s.controlled[1] = 16;
  Object.assign(s.players[16], { x: 0, z: 0 });
  Object.assign(s.ball, {
    x: 0,
    z: 0,
    h: 0.75,
    vx: 1,
    charged: true,
    electric: 1,
    electricBudget: 1,
    lastTouch: 7,
  });
  catchBall(s);
  assert.equal(s.ball.owner, -1);
  assert.equal(s.ball.electric, 0);
  assert.equal(s.ball.charged, true);
  s.players[16].stun = 10;
  s.controlled[0] = 7;
  Object.assign(s.players[7], { x: 0, z: 0 });
  catchBall(s);
  assert.equal(s.ball.owner, 7);
  assert.equal(s.ball.charged, true);
  throwBall(s, 7, false);
  assert.equal(s.ball.charged, false);
  assert.equal(s.ball.electricBudget, 1);
});
test("opposing moving catch clears exhausted charge; stationary catch preserves it", () => {
  for (const vx of [0, 1]) {
    const s = initial();
    s.controlled[1] = 16;
    Object.assign(s.players[16], { x: 0, z: 0 });
    Object.assign(s.ball, {
      x: 0,
      z: 0,
      h: 0.75,
      vx,
      charged: true,
      electric: 0,
      electricBudget: 0,
      lastTouch: 7,
    });
    catchBall(s);
    assert.equal(s.ball.owner, 16);
    assert.equal(s.ball.charged, vx === 0);
  }
});

test("electroball knockdown follows nominal direction at three terrain units", () => {
  const unit = (25 * 22.4) / 576;
  for (const [dx, dz] of [
    [1, 0],
    [1, 1],
    [0, 1],
    [-1, 1],
    [-1, 0],
    [-1, -1],
    [0, -1],
    [1, -1],
  ]) {
    const s = initial();
    s.controlled[1] = 16;
    Object.assign(s.players[16], { x: 0, z: 0 });
    Object.assign(s.players[7], { fx: -dx, fz: -dz });
    Object.assign(s.ball, {
      owner: -1,
      x: 0,
      z: 0,
      h: 0.75,
      dirX: dx,
      dirZ: dz,
      vx: dx * 8 + (dx === 0 ? 4 : 0),
      vz: dz * 8 + (dz === 0 ? 4 : 0),
      charged: true,
      electric: 1,
      electricBudget: 1,
      lastTouch: 7,
    });
    catchBall(s);
    const p = s.players[16];
    assert.equal(p.action, 4);
    assert.equal(p.stun, 35 / 25);
    assert.equal(p.fallX, dx * 3 * unit);
    assert.equal(p.fallZ, dz * 3 * unit);
    assert.equal(s.ball.electric, 0);
    assert.equal(s.ball.owner, -1);
  }
});
